actions.go

```diff
@@ -0,0 +1,123 @@
+	// The function to execute when the action is called
+	Execute func(s *discordgo.Session, m *discordgo.MessageCreate, parameters []string) error
+	// The function to display information about the action
+	DisplayHelp func(s *discordgo.Session, m *discordgo.MessageCreate) error
+}
+
+// MakeExecuteAction type from an Action
+func (action *Action) MakeExecuteAction(parameters []string) ExecuteAction {
+	if len(parameters) == 1 && strings.ToUpper(parameters[0]) == "HELP" {
+		return (*action).DisplayHelp
+	}
+	return func(s *discordgo.Session, m *discordgo.MessageCreate) error {
+		return (*action).Execute(s, m, parameters)
+	}
+}
+
+// GetPermission is used with the security module to check the permission of the action
+func (action *Action) GetPermission() security.Permission {
+	return action.Permission
+}
+
+// ResolveAction looks if the calling parameters match a sub action it calls a ResolveAction on it, and return itself with an ExecuteAction if not.
+// If the parameter is help, return a DisplayHelp ExecuteAction
+func (action *Action) ResolveAction(call []string) (ActionResolver, ExecuteAction) {
+	if (len(call) == 0) || (len(call) == 1 && strings.ToUpper(call[0]) == "HELP") {
+		return action, action.MakeExecuteAction(call)
+	}
+	if subAction, ok := (*action).SubActions[strings.ToUpper(call[0])]; ok {
+		return subAction.ResolveAction(call[1:])
+	}
+	return action, action.MakeExecuteAction(call)
+}
+
+var (
+	// PingAction for Kadok to respond pong
+	PingAction = Action{
+		security.GetCharacterList,
+		map[string]ActionResolver{},
+		func(s *discordgo.Session, m *discordgo.MessageCreate, parameters []string) error {
+			_, err := s.ChannelMessageSend(m.ChannelID, "À Kadoc ! À Kadoc ! Pong!")
+			return err
+		},
+		func(s *discordgo.Session, m *discordgo.MessageCreate) error {
+			return errors.New("Not a valid action")
+		},
+	}
+
+	// PongAction for Kadok to respond ping
+	PongAction = Action{
+		security.GetCharacterList,
+		map[string]ActionResolver{},
+		func(s *discordgo.Session, m *discordgo.MessageCreate, parameters []string) error {
+			_, err := s.ChannelMessageSend(m.ChannelID, "À Kadoc ! À Kadoc ! Ping!")
+			return err
+		},
+		func(s *discordgo.Session, m *discordgo.MessageCreate) error {
+			return errors.New("Not a valid action")
+		},
+	}
+
+	// GetCharactersAction to retrieve the list of available characters
+	GetCharactersAction = Action{
+		security.GetCharacterList,
+		map[string]ActionResolver{},
+		func(s *discordgo.Session, m *discordgo.MessageCreate, parameters []string) error {
+			message := ""
+			message += "\nLe caca des pigeons, c'est caca. Si tu parle d'un de mes amis, je te dirai ce qu'il a dit."
+			message += "\nJ'ai pleins d'amis à Kaamelott:"
+			for _, character := range Configuration.Characters.List {
+				message += "\n- " + character.Name
+			}
+			_, err := s.ChannelMessageSend(m.ChannelID, message)
+			return err
+		},
+		func(s *discordgo.Session, m *discordgo.MessageCreate) error {
```

Comment:
this code is duplicated for almost all actions

Reply:
Following this remark. As most of the actions (if not all) should respond with a message in the chat. To at least give some sort of feedback to the user. The signature of the function type `ExecuteAction` is now:

```golang
type ExecuteAction func(s *discordgo.Session, m *discordgo.MessageCreate) (string, error)
```
The function now returns a string. The responsibility to send it on the channel is given to the caller. This should reduce the replication of this line.

_Note: if the string is equal to `""`, no message is sent._